File: game-runtime/packages/sim/src/geometry/integer-geometry.ts
/**
 * Integer 2D geometry primitives for authoritative map collision.
 *
 * Coordinates are integer millimeters with |value| < 1_000_000. Every
 * product stays far below Number.MAX_SAFE_INTEGER provided segment lengths
 * are capped (see MAX_SEGMENT_LENGTH_MM subdivision in the collision field),
 * so all predicates are exact and deterministic across platforms.
 */

import type { Vec2Mm } from '@jwgb/core';

/** Longest segment the distance query accepts without overflow risk. */
export const MAX_SEGMENT_LENGTH_MM = 32_768;

export function crossOrientation(a: Vec2Mm, b: Vec2Mm, point: Vec2Mm): number {
  return (b.x - a.x) * (point.z - a.z) - (b.z - a.z) * (point.x - a.x);
}

function isPointOnSegment(a: Vec2Mm, b: Vec2Mm, point: Vec2Mm): boolean {
  return (
    Math.min(a.x, b.x) <= point.x &&
    point.x <= Math.max(a.x, b.x) &&
    Math.min(a.z, b.z) <= point.z &&
    point.z <= Math.max(a.z, b.z)
  );
}

/** Inclusive integer segment intersection used by line-of-sight queries. */
export function segmentsIntersect(
  firstStart: Vec2Mm,
  firstEnd: Vec2Mm,
  secondStart: Vec2Mm,
  secondEnd: Vec2Mm,
): boolean {
  const firstToSecondStart = crossOrientation(firstStart, firstEnd, secondStart);
  const firstToSecondEnd = crossOrientation(firstStart, firstEnd, secondEnd);
  const secondToFirstStart = crossOrientation(secondStart, secondEnd, firstStart);
  const secondToFirstEnd = crossOrientation(secondStart, secondEnd, firstEnd);
  const firstStraddles =
    (firstToSecondStart < 0 && firstToSecondEnd > 0) ||
    (firstToSecondStart > 0 && firstToSecondEnd < 0);
  const secondStraddles =
    (secondToFirstStart < 0 && secondToFirstEnd > 0) ||
    (secondToFirstStart > 0 && secondToFirstEnd < 0);
  if (firstStraddles && secondStraddles) {
    return true;
  }
  return (
    (firstToSecondStart === 0 && isPointOnSegment(firstStart, firstEnd, secondStart)) ||
    (firstToSecondEnd === 0 && isPointOnSegment(firstStart, firstEnd, secondEnd)) ||
    (secondToFirstStart === 0 && isPointOnSegment(secondStart, secondEnd, firstStart)) ||
    (secondToFirstEnd === 0 && isPointOnSegment(secondStart, secondEnd, firstEnd))
  );
}

export function distanceSquaredToSegment(point: Vec2Mm, start: Vec2Mm, end: Vec2Mm): number {
  const deltaX = end.x - start.x;
  const deltaZ = end.z - start.z;
  const lengthSquared = deltaX * deltaX + deltaZ * deltaZ;
  if (lengthSquared === 0) {
    return distanceSquaredBetween(point, start);
  }
  const projection = Math.max(
    0,
    Math.min(lengthSquared, (point.x - start.x) * deltaX + (point.z - start.z) * deltaZ),
  );
  const closest = {
    x: start.x + Math.trunc((deltaX * projection) / lengthSquared),
    z: start.z + Math.trunc((deltaZ * projection) / lengthSquared),
  };
  return distanceSquaredBetween(point, closest);
}

/** Exact even-odd containment for a possibly concave ring. */
export function ringContainsPoint(ring: readonly Vec2Mm[], point: Vec2Mm): boolean {
  let inside = false;
  for (let index = 0; index < ring.length; index += 1) {
    const a = ring[index] as Vec2Mm;
    const b = ring[(index + 1) % ring.length] as Vec2Mm;
    if (a.z > point.z === b.z > point.z) {
      continue;
    }
    // point.x < a.x + (point.z - a.z) * (b.x - a.x) / (b.z - a.z), cross-multiplied.
    const deltaZ = b.z - a.z;
    const left = (point.x - a.x) * deltaZ;
    const right = (point.z - a.z) * (b.x - a.x);
    if (deltaZ > 0 ? left < right : left > right) {
      inside = !inside;
    }
  }
  return inside;
}

/** Exact containment for a convex CCW ring, boundary inclusive. */
export function convexContainsPoint(vertices: readonly Vec2Mm[], point: Vec2Mm): boolean {
  for (let index = 0; index < vertices.length; index += 1) {
    const a = vertices[index] as Vec2Mm;
    const b = vertices[(index + 1) % vertices.length] as Vec2Mm;
    if (crossOrientation(a, b, point) < 0) {
      return false;
    }
  }
  return true;
}

/**
 * Closest lattice point on segment [a, b] to the query point.
 *
 * The projection parameter is truncated to integers, so the result can be
 * up to 1 mm away from the mathematically exact foot point. That epsilon is
 * acceptable for wall padding and keeps every intermediate value integral.
 * Requires |b - a| <= MAX_SEGMENT_LENGTH_MM.
 */
export function closestPointOnSegment(a: Vec2Mm, b: Vec2Mm, point: Vec2Mm): Vec2Mm {
  const abX = b.x - a.x;
  const abZ = b.z - a.z;
  const lengthSquared = abX * abX + abZ * abZ;
  if (lengthSquared === 0) {
    return a;
  }
  const apX = point.x - a.x;
  const apZ = point.z - a.z;
  const rawT = apX * abX + apZ * abZ;
  if (rawT <= 0) {
    return a;
  }
  if (rawT >= lengthSquared) {
    return b;
  }
  return {
    x: a.x + Math.trunc((abX * rawT) / lengthSquared),
    z: a.z + Math.trunc((abZ * rawT) / lengthSquared),
  };
}

export function distanceSquaredBetween(a: Vec2Mm, b: Vec2Mm): number {
  const dx = a.x - b.x;
  const dz = a.z - b.z;
  return dx * dx + dz * dz;
}

/**
 * True when a circle overlaps a convex CCW polygon whose edges are at most
 * MAX_SEGMENT_LENGTH_MM long (enforced by the collision field constructor).
 */
export function circleOverlapsConvex(
  vertices: readonly Vec2Mm[],
  center: Vec2Mm,
  radiusMm: number,
): boolean {
  if (convexContainsPoint(vertices, center)) {
    return true;
  }
  const radiusSquared = radiusMm * radiusMm;
  for (let index = 0; index < vertices.length; index += 1) {
    const a = vertices[index] as Vec2Mm;
    const b = vertices[(index + 1) % vertices.length] as Vec2Mm;
    const closest = closestPointOnSegment(a, b, center);
    if (distanceSquaredBetween(closest, center) <= radiusSquared) {
      return true;
    }
  }
  return false;
}

/** Splits a segment into deterministic chunks no longer than the cap. */
export function subdivideSegment(a: Vec2Mm, b: Vec2Mm, maximumLengthMm: number): Vec2Mm[][] {
  const spanX = b.x - a.x;
  const spanZ = b.z - a.z;
  const span = Math.max(Math.abs(spanX), Math.abs(spanZ));
  const chunkCount = Math.max(1, Math.ceil(span / maximumLengthMm));
  const segments: Vec2Mm[][] = [];
  let previous = a;
  for (let chunk = 1; chunk <= chunkCount; chunk += 1) {
    const next =
      chunk === chunkCount
        ? b
        : {
            x: a.x + Math.trunc((spanX * chunk) / chunkCount),
            z: a.z + Math.trunc((spanZ * chunk) / chunkCount),
          };
    segments.push([previous, next]);
    previous = next;
  }
  return segments;
}

File: game-runtime/unity/Packages/com.jwgb.sim/Runtime/Deterministic/Geometry/IntegerGeometry.cs
using Jwgb.Content;

namespace Jwgb.Sim.Deterministic
{
    /// <summary>
    /// Integer 2D geometry primitives for authoritative map collision.
    /// Mirrors packages/sim/src/geometry/integer-geometry.ts exactly; every
    /// product stays inside long range because segment lengths are capped by
    /// <see cref="MaxSegmentLengthMm"/> subdivision.
    /// </summary>
    internal static class IntegerGeometry
    {
        public const long MaxSegmentLengthMm = 32_768;

        public static long CrossOrientation(
            MapPointMmRecord a,
            MapPointMmRecord b,
            MapPointMmRecord point)
        {
            return ((b.X - a.X) * (point.Z - a.Z)) - ((b.Z - a.Z) * (point.X - a.X));
        }

        public static bool RingContainsPoint(
            MapPointMmRecord[] ring,
            MapPointMmRecord point)
        {
            var inside = false;
            for (var index = 0; index < ring.Length; index += 1)
            {
                var a = ring[index];
                var b = ring[(index + 1) % ring.Length];
                if ((a.Z > point.Z) == (b.Z > point.Z))
                {
                    continue;
                }

                var deltaZ = b.Z - a.Z;
                var left = (point.X - a.X) * deltaZ;
                var right = (point.Z - a.Z) * (b.X - a.X);
                if (deltaZ > 0 ? left < right : left > right)
                {
                    inside = !inside;
                }
            }

            return inside;
        }

        public static bool ConvexContainsPoint(
            MapPointMmRecord[] vertices,
            MapPointMmRecord point)
        {
            for (var index = 0; index < vertices.Length; index += 1)
            {
                var a = vertices[index];
                var b = vertices[(index + 1) % vertices.Length];
                if (CrossOrientation(a, b, point) < 0)
                {
                    return false;
                }
            }

            return true;
        }

        /// <summary>
        /// Closest lattice point on [a, b]; up to 1 mm from the exact foot
        /// point, matching the TypeScript truncation semantics.
        /// </summary>
        public static MapPointMmRecord ClosestPointOnSegment(
            MapPointMmRecord a,
            MapPointMmRecord b,
            MapPointMmRecord point)
        {
            var abX = b.X - a.X;
            var abZ = b.Z - a.Z;
            var lengthSquared = (abX * abX) + (abZ * abZ);
            if (lengthSquared == 0)
            {
                return a;
            }

            var apX = point.X - a.X;
            var apZ = point.Z - a.Z;
            var rawT = (apX * abX) + (apZ * abZ);
            if (rawT <= 0)
            {
                return a;
            }

            if (rawT >= lengthSquared)
            {
                return b;
            }

            // C# integer division truncates toward zero exactly like
            // Math.trunc on the float64-exact products used in TypeScript.
            return new MapPointMmRecord(
                a.X + ((abX * rawT) / lengthSquared),
                a.Z + ((abZ * rawT) / lengthSquared));
        }

        public static long DistanceSquaredBetween(
            MapPointMmRecord a,
            MapPointMmRecord b)
        {
            var dx = a.X - b.X;
            var dz = a.Z - b.Z;
            return (dx * dx) + (dz * dz);
        }

        private static bool IsPointOnSegment(
            MapPointMmRecord a,
            MapPointMmRecord b,
            MapPointMmRecord point)
        {
            return System.Math.Min(a.X, b.X) <= point.X &&
                point.X <= System.Math.Max(a.X, b.X) &&
                System.Math.Min(a.Z, b.Z) <= point.Z &&
                point.Z <= System.Math.Max(a.Z, b.Z);
        }

        /// <summary>
        /// Inclusive integer segment intersection used by line-of-sight
        /// queries. Port of segmentsIntersect from
        /// packages/sim/src/geometry/integer-geometry.ts.
        /// </summary>
        public static bool SegmentsIntersect(
            MapPointMmRecord firstStart,
            MapPointMmRecord firstEnd,
            MapPointMmRecord secondStart,
            MapPointMmRecord secondEnd)
        {
            var firstToSecondStart = CrossOrientation(
                firstStart,
                firstEnd,
                secondStart);
            var firstToSecondEnd = CrossOrientation(
                firstStart,
                firstEnd,
                secondEnd);
            var secondToFirstStart = CrossOrientation(
                secondStart,
                secondEnd,
                firstStart);
            var secondToFirstEnd = CrossOrientation(
                secondStart,
                secondEnd,
                firstEnd);
            var firstStraddles =
                (firstToSecondStart < 0 && firstToSecondEnd > 0) ||
                (firstToSecondStart > 0 && firstToSecondEnd < 0);
            var secondStraddles =
                (secondToFirstStart < 0 && secondToFirstEnd > 0) ||
                (secondToFirstStart > 0 && secondToFirstEnd < 0);
            if (firstStraddles && secondStraddles)
            {
                return true;
            }

            return (firstToSecondStart == 0 &&
                    IsPointOnSegment(firstStart, firstEnd, secondStart)) ||
                (firstToSecondEnd == 0 &&
                    IsPointOnSegment(firstStart, firstEnd, secondEnd)) ||
                (secondToFirstStart == 0 &&
                    IsPointOnSegment(secondStart, secondEnd, firstStart)) ||
                (secondToFirstEnd == 0 &&
                    IsPointOnSegment(secondStart, secondEnd, firstEnd));
        }

        /// <summary>
        /// Port of distanceSquaredToSegment from
        /// packages/sim/src/geometry/integer-geometry.ts.
        /// </summary>
        public static long DistanceSquaredToSegment(
            MapPointMmRecord point,
            MapPointMmRecord start,
            MapPointMmRecord end)
        {
            var deltaX = end.X - start.X;
            var deltaZ = end.Z - start.Z;
            var lengthSquared = (deltaX * deltaX) + (deltaZ * deltaZ);
            if (lengthSquared == 0)
            {
                return DistanceSquaredBetween(point, start);
            }

            var rawProjection =
                ((point.X - start.X) * deltaX) +
                ((point.Z - start.Z) * deltaZ);
            var projection = System.Math.Max(
                0,
                System.Math.Min(lengthSquared, rawProjection));
            var closest = new MapPointMmRecord(
                start.X + ((deltaX * projection) / lengthSquared),
                start.Z + ((deltaZ * projection) / lengthSquared));
            return DistanceSquaredBetween(point, closest);
        }

        public static int SubdivideChunkCount(
            MapPointMmRecord a,
            MapPointMmRecord b,
            long maximumLengthMm)
        {
            var spanX = b.X - a.X;
            var spanZ = b.Z - a.Z;
            var span = System.Math.Max(System.Math.Abs(spanX), System.Math.Abs(spanZ));
            var chunkCount = (span + maximumLengthMm - 1) / maximumLengthMm;
            return (int)System.Math.Max(1, chunkCount);
        }

        public static MapPointMmRecord SubdividePoint(
            MapPointMmRecord a,
            MapPointMmRecord b,
            int chunk,
            int chunkCount)
        {
            if (chunk >= chunkCount)
            {
                return b;
            }

            var spanX = b.X - a.X;
            var spanZ = b.Z - a.Z;
            return new MapPointMmRecord(
                a.X + ((spanX * chunk) / chunkCount),
                a.Z + ((spanZ * chunk) / chunkCount));
        }
    }
}
